extensions/Worker.Extensions.Shared/Extensions/TypeExtensions.cs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Linq;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions
+{
+    internal static class TypeExtensions
+    {
+        /// <summary>
+        /// Returns true if the target type is a collection type.
+        /// </summary>
+        public static bool IsCollectionType(this Type type)
+        {
+            if (type is null)
+            {
+                throw new ArgumentNullException(nameof(type));
+            }
+
+            // Edge case: string implements IEnumerable<char> and byte[] would pass the IsArray check
+            if (type == typeof(string) || type == typeof(byte[]))
+            {
+                return false;
+            }
+
+            if (!(type.IsArray || typeof(IEnumerable).IsAssignableFrom(type)))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Returns the element type of an collection.
+        /// </summary>
+        public static bool TryGetCollectionElementType(this Type type, out Type? elementType)
+        {
+            elementType = null;
+
+            if (!type.IsCollectionType())
+            {
+                return false;
+            }
+
+            // Check if the type is an array
+            if (type.IsArray)
+            {
+                elementType = type.GetElementType();
+                return true;
+            }
+
+            // Traverse the inheritance hierarchy to find the first generic interface
+            while (type != typeof(object) || type is not null)
+            {
+                var interfaceType = type.GetInterfaces().FirstOrDefault(t => t.IsGenericType
```

Comment:
You're right, it returns `typeof(int)` - what would be the best way to approach this specific scenario? Moving the `IsGeneric` check down and having a split branch i.e.

```csharp
var interfaceType = type.GetInterfaces().FirstOrDefault((t == typeof(IEnumerable) || t == typeof(ICollection) || t == typeof(IList)));

if (interfaceType is not null)
{
    if (interfaceType.IsGenericType)
    {
        // do a?
    }
    else
    {
        // do b?
    }
}
```

Reply:
No that doesn't make sense either - I'm stumped

Reply:
You can do something like: if generic, get generic type definition, check if that is `IEnumerable<>`, `ICollection<>` or `IList<>`. If so, then pull off the first generic argument.

https://learn.microsoft.com/en-us/dotnet/api/system.type.getgenerictypedefinition?view=net-7.0

Reply:
Fixed! I realize now why I had a hard time with this, I didn't have "using System.Collections.Generic;" and the error message was not clear as to why `== typeof(IEnumerable<>)` was not okay 😔 Anywho, this case now correctly gets string instead of int